### PR TITLE
Add "Initialize with a default value" quick fix

### PR DESCRIPTION
**Pull request by ortem**
Add the quick fix for "Use of uninitialized variable" error. Includes the refactoring of `AddStructFieldsFix.kt`.
Merge after #3315

### CODE REVIEW

**Comment by amakeev** [comment]
The name of the fix is somewhat ambiguous. Maybe 'Initialize with a default value'?

**Comment by vlad20012** [comment]
image
(it's better to recover the state from the element)

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        val initExprNavigationElement = newDeclaration.expr?.navigationElement as? NavigatablePsiElement ?: return
+        initExprNavigationElement.navigate(true)
+    }
+
+    companion object {
+        fun createIfCompatible(element: RsElement): InitializeWithDefaultValueFix? {
+            val variable = element.ancestorOrSelf<RsExpr>() ?: return null
+            val patBinding = variable.declaration as? RsPatBinding ?: return null
+            val declaration = patBinding.ancestorOrSelf<RsLetDecl>() ?: return null
+            val typeReference = declaration.typeReference ?: return null
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: Let's use `patBinding.type` to cover a case like
```rust
fn main() {
    let x;
    let y: i32 = x;
}
```

File: src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+class RsDefaultValueBuilder(
+    val items: KnownItems,
+    val mod: RsMod,
+    val psiFactory: RsPsiFactory,
+    val recursive: Boolean = true,
+    val structFiller: RsStructFiller = RsStructFiller(items, mod, psiFactory, recursive)
+) {
+    fun buildFor(ty: Ty): RsExpr {
+        val defaultValue = { psiFactory.createExpression("()") }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt]: Can now be a private method of the class

File: src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+package org.rust.lang.core.psi
+
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.KnownItems
+import org.rust.lang.core.types.ty.*
+import org.rust.lang.core.types.type
+
+class RsDefaultValueBuilder(
+    val items: KnownItems,
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt]: All properties can be private

File: src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+package org.rust.lang.core.psi
+
+import org.rust.lang.core.psi.ext.RsFieldDecl
+import org.rust.lang.core.psi.ext.RsMod
+import org.rust.lang.core.resolve.KnownItems
+import org.rust.lang.core.types.type
+
+class RsStructFiller(val items: KnownItems, val mod: RsMod, val psiFactory: RsPsiFactory, val recursive: Boolean) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt]: All properties can be private

File: src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            return existingFields.getOrNull(prevIdx + 1)
+        }
+
+        return null
+    }
+
+    private fun specializedCreateStructLiteralField(fieldDecl: RsFieldDecl): RsStructLiteralField? {
+        val fieldName = fieldDecl.name ?: return null
+        val fieldType = fieldDecl.typeReference?.type ?: return null
+        val fieldLiteral = RsDefaultValueBuilder(items, mod, psiFactory, recursive).buildFor(fieldType)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt]: `items, mod, psiFactory, recursive, this`?
Can also be a field of the current class?

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    private val declaration: RsLetDecl,
+    private val variableName: String,
+    private val typeReference: RsTypeReference
+) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText() = "Initialize with a default value"
+    override fun getFamilyName() = name
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        val psiFactory = RsPsiFactory(project)
+        val defaultValueFactory = RsDefaultValueBuilder(declaration.knownItems, declaration.containingMod, psiFactory)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: I don't sure that we should use recursive fill here

File: src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt (hunk 4 above)
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt]: I don't like this 2 class that refers to each other recursive call to each other and containing of the same properties. Maybe single class is enough?

Comment (ortem) [reply src/main/kotlin/org/rust/lang/core/psi/RsStructFiller.kt]: OK, now I combined it

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt (hunk 6 above)
Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: Now `recursive: Boolean = false` by default

**Comment by vlad20012** [comment]
This intention always adds explicit type to let statement:
peek 2019-02-23 16-08

And also removes statement attributes:
peek 2019-02-23 16-10

I guess there are also a lot of problems with comments. I think we should keep let statement in PSI as is and just add initializer expression to it.

File: src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+import org.rust.lang.core.types.type
+
+class RsDefaultValueBuilder(
+    private val items: KnownItems,
+    private val mod: RsMod,
+    private val psiFactory: RsPsiFactory,
+    private val recursive: Boolean = false
+) {
+    private val defaultValue: RsExpr
+        get() = psiFactory.createExpression("()")
```
Comment (ice1000) [inline src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt]: How about `Default::default()`?

**Comment by ortem** [comment]
@vlad20012 Done

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        val psiFactory = RsPsiFactory(project)
+        val initExpr = RsDefaultValueBuilder(declaration.knownItems, declaration.containingMod, psiFactory).buildFor(patBinding.type)
+        declaration.addBefore(psiFactory.createEq(), semicolon)
+        declaration.addBefore(initExpr, semicolon)
+        if (editor != null) {
+            val declarationExpr = declaration.expr ?: return
+            val documentManager = PsiDocumentManager.getInstance(project)
+            documentManager.commitDocument(editor.document)
+            documentManager.doPostponedOperationsAndUnblockDocument(editor.document)
+            editor.caretModel.moveToOffset(declarationExpr.textOffset)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: Let's place the caret __after__ inserted expression to make it simple to remove it

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt [hunk 9]
```diff
@@ -131,8 +132,8 @@ class RsPsiFactory(private val project: Project, private val markGenerated: Bool
         createFromText("fn main() { $text 92; }")
             ?: error("Failed to create statement from text: `$text`")
 
-    fun createLetDeclaration(name: String, expr: RsExpr, mutable: Boolean = false, type: RsTypeReference? = null): RsLetDecl =
-        createStatement("let ${"mut".iff(mutable)}$name${if (type != null) ": ${type.text}" else ""} = ${expr.text};") as RsLetDecl
+    fun createLetDeclaration(name: String, expr: RsExpr, mutable: Boolean = false, type: Ty? = null): RsLetDecl =
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt]: These changes aren't needed now?

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    override fun getFamilyName() = name
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        val variable = startElement.ancestorOrSelf<RsExpr>() ?: return
+        val patBinding = variable.declaration as? RsPatBinding ?: return
+        val declaration = patBinding.ancestorOrSelf<RsLetDecl>() ?: return
+        val semicolon = declaration.semicolon ?: return
+        val psiFactory = RsPsiFactory(project)
+        val initExpr = RsDefaultValueBuilder(declaration.knownItems, declaration.containingMod, psiFactory).buildFor(patBinding.type)
+        declaration.addBefore(psiFactory.createEq(), semicolon)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: What if EQ exists already?

File: src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+class InitializeWithDefaultValueFix(element: RsElement) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
+    override fun getText() = "Initialize with a default value"
+    override fun getFamilyName() = name
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        val variable = startElement.ancestorOrSelf<RsExpr>() ?: return
+        val patBinding = variable.declaration as? RsPatBinding ?: return
+        val declaration = patBinding.ancestorOrSelf<RsLetDecl>() ?: return
+        val semicolon = declaration.semicolon ?: return
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt]: If there is no semicolon, quickfix is available, but nothing happens if I try to invoke it

File: src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt (hunk 7 above)
Comment (ice1000) [reply src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt]: Hmm, works for me

**Pull request merged**